storage/index_insert_queue.go

```diff
@@ -0,0 +1,258 @@
+) databaseIndexInsertQueue {
+	currBatch := &dbIndexInsertBatch{}
+	currBatch.Reset()
+	return &dbIndexInsertQueue{
+		currBatch:           currBatch,
+		indexBatchBackoff:   defaultIndexBatchBackoff,
+		indexPerSecondLimit: defaultIndexPerSecondLimit,
+		indexBatchFn:        indexBatchFn,
+		nowFn:               nowFn,
+		sleepFn:             time.Sleep,
+		notifyInsert:        make(chan struct{}, 1),
+		metrics:             newDatabaseIndexInsertQueueMetrics(scope),
+	}
+}
+
+func (q *dbIndexInsertQueue) insertLoop() {
+	var lastInsert time.Time
+	freeBatch := &dbIndexInsertBatch{}
+	freeBatch.Reset()
+	for range q.notifyInsert {
+		// Check if inserting too fast
+		elapsedSinceLastInsert := q.nowFn().Sub(lastInsert)
+
+		// Rotate batches
+		var (
+			state   dbIndexInsertQueueState
+			backoff time.Duration
+			batch   *dbIndexInsertBatch
+		)
+		q.Lock()
+		state = q.state
+		if elapsedSinceLastInsert < q.indexBatchBackoff {
+			// Need to backoff before rotate and insert
+			backoff = q.indexBatchBackoff - elapsedSinceLastInsert
+		} else {
+			// No backoff required, rotate and go
+			batch = q.currBatch
+			q.currBatch = freeBatch
+		}
+		q.Unlock()
+
+		if backoff > 0 {
+			q.sleepFn(backoff)
+			q.Lock()
+			// Rotate after backoff
+			batch = q.currBatch
+			q.currBatch = freeBatch
+			q.Unlock()
+		}
+
+		if len(batch.inserts) > 0 {
+			q.indexBatchFn(batch.inserts)
+		}
+		batch.wg.Done()
+
+		// Set the free batch
+		batch.Reset()
+		freeBatch = batch
+
+		lastInsert = q.nowFn()
+
+		if state != dbIndexInsertQueueStateOpen {
+			return // Break if the queue closed
+		}
+	}
+}
+
+func (q *dbIndexInsertQueue) Insert(
+	d doc.Document,
+	fns indexInsertLifecycleHooks,
+) (*sync.WaitGroup, error) {
+	windowNanos := q.nowFn().Truncate(time.Second).UnixNano()
+
+	q.Lock()
+	if q.state != dbIndexInsertQueueStateOpen {
+		q.Unlock()
+		return nil, errIndexInsertQueueNotOpen
+	}
+	if limit := q.indexPerSecondLimit; limit > 0 {
+		if q.indexPerSecondLimitWindowNanos != windowNanos {
+			// Rolled into to a new window
+			q.indexPerSecondLimitWindowNanos = windowNanos
+			q.indexPerSecondLimitWindowValues = 0
+		}
+		q.indexPerSecondLimitWindowValues++
+		if q.indexPerSecondLimitWindowValues > limit {
+			q.Unlock()
+			return nil, errNewSeriesIndexRateLimitExceeded
+		}
+	}
+	q.currBatch.inserts = append(q.currBatch.inserts, dbIndexInsert{
+		doc: d,
+		fns: fns,
+	})
+	wg := q.currBatch.wg
+	q.Unlock()
+
+	// Notify insert loop
+	select {
+	case q.notifyInsert <- struct{}{}:
+	default:
+		// Loop busy, already ready to consume notification
+	}
+
+	q.metrics.numPending.Inc(1)
+	return wg, nil
+}
+
+func (q *dbIndexInsertQueue) Start() error {
+	q.Lock()
+	defer q.Unlock()
+
+	if q.state != dbIndexInsertQueueStateNotOpen {
+		return errIndexInsertQueueAlreadyOpenOrClosed
+	}
+
+	q.state = dbIndexInsertQueueStateOpen
+	go q.insertLoop()
+	return nil
+}
+
+func (q *dbIndexInsertQueue) Stop() error {
+	q.Lock()
+	defer q.Unlock()
```

Comment:
Do you want to hold the lock for the whole duration here? Seems like you'd want to:

1. Grab the lock
2. Mark the insert queue as closed
3. Release the lock

That way callers that are trying to call Write() can receive an immediate error instead of waiting for a while on the lock only to be told the insert queue is closed

Reply:
There's no blocking in the code below, there's is a best effort to kick off the final flush, so whether you release it slightly earlier there will be no real difference in how fast this method executes.

Reply:
The latency spike during the close path is such an edge case, don't need think about it while there's lots of core functionality to be developed.

Reply:
fair enough